src/object/list.rs

```diff
@@ -270,3 +272,53 @@ impl List {
         self.len() == 0
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use crate::{r, r_expect};
+    #[test]
+    fn list_declaration_ambiguity() {
+        assert_eq!(r!((a = 1,)), r!(list(a = 1)));
+        assert_eq!(r!((a = 1)), r!(1));
+        assert_eq!(r!((1)), r!(1));
+        assert_eq!(r!((1,)), r!(list(1)));
+    }
+
+    #[test]
+    fn copy_on_write_single_bracket() {
+        r_expect! {{"
+            l1 = (1,)
+            l2 = l1
+            l1[1] = 2
+            l1[[1]] == 2 & l2[[1]] == 1
+        "}}
+    }
+    #[test]
+    fn copy_on_write_bracket_names() {
+        r_expect! {{r#"
+            l1 = (a = 1,)
+            l2 = l1
+            l1["a"] = 2
+            l1$a == 2 & l2$a == 1
+        "#}}
+    }
+    #[test]
+    fn copy_on_write_slice_names() {
+        r_expect! {{r#"
+            l = (a = 1, b = 2, c = 3)
+            l1 = l
+            l1[c("a", "b")] = c(10, 20)
+
+            l1$a == 10 && l1$b == 20 & l$a == 1 & l$b == 2
+        "#}}
+    }
+    #[test]
+    fn copy_on_write_slice_indices() {
+        r_expect! {{"
+            l = (1, 2)
+            l1 = l
+            l1[1:2] = [10, 20]
+            l1[[1]] == 10 && l1[[2]] == 20 & l[[1]] == 1 & l[[2]] == 2
+        "}}
+    }
+}
```

Comment:
Great catch! I think this means that for lists, we might have to change how the `.clone()` method behaves / need a slight variant of the `Data<T>` construct.

Currently, when we clone a `VecData<(Option<String>, Obj))` (i.e. a `Rc<RefCell<Rc<vec![(Option<String>, Obj)]`), the inner `Obj` is not cloned. 
This means that when we do 

```r
l1 = (1, )
l2 = l2
```

`l1` and `l2` share the same `Obj` that represents the `1`.

If we then do `l2[[1]] = 1`, this will first get a reference into the inner Object (`&Obj`) and then call `.assign` on it, directly mutating the object that is contained both in `l1` and `l2`.

I will try to figure out a solution tomorrow.

Reply:
Nevermind, the solution still works.